x509-cert/src/certificate.rs

```diff
@@ -84,29 +85,29 @@ pub struct TbsCertificate<'a> {
     pub version: Version,
 
     pub serial_number: SerialNumber,
-    pub signature: AlgorithmIdentifierRef<'a>,
+    pub signature: AlgorithmIdentifier<SignParams>,
     pub issuer: Name,
     pub validity: Validity,
     pub subject: Name,
-    pub subject_public_key_info: SubjectPublicKeyInfoRef<'a>,
+    pub subject_public_key_info: SubjectPublicKeyInfo<KeyParams, Key>,
 
     #[asn1(context_specific = "1", tag_mode = "IMPLICIT", optional = "true")]
-    pub issuer_unique_id: Option<BitStringRef<'a>>,
+    pub issuer_unique_id: Option<BitString>,
 
     #[asn1(context_specific = "2", tag_mode = "IMPLICIT", optional = "true")]
-    pub subject_unique_id: Option<BitStringRef<'a>>,
+    pub subject_unique_id: Option<BitString>,
 
     #[asn1(context_specific = "3", tag_mode = "EXPLICIT", optional = "true")]
-    pub extensions: Option<crate::ext::Extensions<'a>>,
+    pub extensions: Option<crate::ext::Extensions>,
 }
 
-impl<'a> TbsCertificate<'a> {
+impl<SignParams, KeyParams, Key> TbsCertificate<SignParams, KeyParams, Key> {
```

Comment:
Rather than making these generic, I think they can just be fixed to `Any`, `Any`, and `BitString` respectively, at least for now.

That will give you a fully owned representation of `TbsCertificate` and `Certificate` which can also handle any kind of signature algorithm, without the additional complexity of generics and variadic representations.

If you really do want to make it generic, I would suggest adding something like `pub trait Profile` to be impl'd on ZSTs which can act as a carrier for the generic parameters.

Reply:
Let's go with fully own for now, but I'll resubmit another PR to follow up on the profile idea.
I think it will ease things a lot on the derive as I can encode bounds on the trait instead. It could very much help with bringing non-owned extensions back as well.